rfcs/33-apps_own_scheduling.md

```diff
@@ -0,0 +1,239 @@
+
+## User-facing explanation
+
+**Plugins** are cohesive, drop-in units of functionality that you can add to your Bevy app.
+Plugins can initialize **resources** to store data to, or add **system sets**, used to add coherent batches of logic to your app.
+
+Plugins can be added to your app using the `App::add_plugin` method.
+If you're new to Bevy, third-party plugins added in this way should immediately work without further configuration.
+
+```rust
+use bevy::prelude::*;
+use bevy_hypothetical_tui::prelude::*;
+
+fn main(){
+  App::minimal().add_plugin(TuiPlugin).run();
+}
+```
+
+You can write your own plugins by exporting a function that returns a `Plugin` struct with the desired system sets and resources.
+This can be provided as a standalone function or, more commonly, as a method on a simple `MyPlugin` struct:
+
+```rust
+struct CombatPlugin;
+
+impl IntoPlugin for CombatPlugin {
+ fn into() -> Plugin {
+    Plugin::default()
+      .add_systems("Combat", SystemSet::new().with(attack_system).with(death_system));
+      .add_system("PlayerMovement", player_movement_system)
+      .init_resouce::<Score>()
+  }
+}
+
+fn main(){
+  App::default().add_plugin(CombatPlugin).run();
+}
+
+```
+
+Under the hood, plugins are very simple, and their systems and resources can be created and configured directly.
+
+The `Plugin` struct stores data in a straightforward way:
+
+```rust
+pub struct Plugin {
+  pub systems: HashMap<Box<dyn SystemLabel>, SystemSet>,
+  pub resources: HashSet<Box<dyn Resource>>,
+}
+```
+
+When you're creating Bevy apps, you can read, add, remove and configure both systems and resources stored within a plugin that you're using, allowing you to fit it into your app's needs using the standard `HashMap` and `HashSet` APIs.
+The methods shown above (such as `.add_system`) are simply convenience methods that manipulate these public fields.
+
+Note that each system set that you include requires its own label to allow other plugins to control their order relative to that system set, and allows the central `App` to further configure them as a unit if needed.
+
+### Hard system ordering rules
+
+In addition to `.before` and `.after`, `.hard_before` and `.hard_after` can be used to specify that two systems must be separated by a hard sync point in the schedule (generally the end of the stage), in addition to following the standard ordering constraint.
+
+This is critical when `Commands` from the first system must complete (e.g. spawning or despawning entities) before the second system can validly run.
+
+Note that hard system ordering rules have no special schedule-altering powers!
+They are merely an assertion to be checked at the time of schedule construction.
+Their purpose is to ensure that you (or the users of your plugin) do not accidentally break the required logic by shuffling when systems run relative to each other.
+
+### Configuring plugins
```

Comment:
I think we should have specific examples of "user-configured plugins" that cover the various scenarios mentioned in the motivation section. What does it look like for a user to adapt the default plugins (ex: transforms) to account for their custom RunCriteria / Stages / States? We should try to cover a few "common" examples, then try to poke holes in them / identify places where assumptions are broken.